### PR TITLE
Adds OAuth user role validation

### PR DESCRIPTION
**Pull request by shubhamparikh927**
This will enable fhir server for role based token validation of oauth user.
It requires below things to pass from environment
     - OAUTH_USER_ROLE (user role which has permission)
     - OAUTH_CLIENT_ID (registered client id in oauth server)

### CODE REVIEW

File: src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java [hunk 1]
```diff
@@ -91,7 +95,10 @@
 			publicKey = StringUtils.isEmpty(publicKey) ? oAuth2Helper.getJwtPublicKey(kid, OAUTH_URL) : publicKey;
 			JWTVerifier verifier = oAuth2Helper.getJWTVerifier(jwt, publicKey);
 			jwt = verifier.verify(token);
-			return allowAll();
+			ArrayList<String> roles = oAuth2Helper.getRoles(jwt);
+			if(!roles.isEmpty() && !Collections.disjoint(roles, ROLES)) {
```
Comment (hankwallace) [inline src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java]: I think the code needs to be reworked. The expectation is certain operations will require a specific role.

Example use cases:
* As a fhir user, I can execute certain queries
* As a fhir admin, I can execute all queries and use the web ui

For now, we only need to support that first use case. In the future when we allow the web ui to be used we might implement the 2nd.

Let's use a single value in the env var.

I would probably be better to pass the required role as an argument to this function. That way the caller can specify the role required for the operation.

Also, it's a minor thing, but there should be a space between `if` and the `(` in code.

**Review by hankwallace — CHANGES_REQUESTED** [review]

File: src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java (hunk 1 above)
Comment (shubhamparikh927) [reply src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java]: ok I will update the changes

File: src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java [hunk 2]
```diff
@@ -29,6 +29,7 @@
 	private static final String APIKEY_HEADER = "x-api-key";
 	private static final String APIKEY = System.getenv("APIKEY");
 	private static final String TOKEN_PREFIX = "BEARER ";
+	private static final String ROLE = System.getenv("OAUTH_USER_ROLE");
```
Comment (hankwallace) [inline src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java]: Consider renaming this var to `OAUTH_USER_ROLE`

File: src/main/java/ca/uhn/fhir/jpa/starter/OAuth2Helper.java [hunk 3]
```diff
@@ -21,10 +23,12 @@
 import com.auth0.jwt.JWTVerifier;
 import com.auth0.jwt.algorithms.Algorithm;
 import com.auth0.jwt.exceptions.JWTVerificationException;
+import com.auth0.jwt.interfaces.Claim;
 import com.auth0.jwt.interfaces.DecodedJWT;
 
 public class OAuth2Helper {
 	private static final Logger logger = LoggerFactory.getLogger(OAuth2Helper.class);
+	private static final String CLIENT_NAME = System.getenv("OAUTH_CLIENT_NAME");
```
Comment (hankwallace) [inline src/main/java/ca/uhn/fhir/jpa/starter/OAuth2Helper.java]: Like the other env vars, I think this should be moved to the interceptor and have the value passed into functions that need it.

Comment (hankwallace) [reply src/main/java/ca/uhn/fhir/jpa/starter/OAuth2Helper.java]: The var should be named `CLIENT_ID`

File: src/main/java/ca/uhn/fhir/jpa/starter/OAuth2Helper.java [hunk 4]
```diff
@@ -104,5 +108,13 @@ protected PublicKey getJwtPublicKey(String kid, String oauthUrl) {
 
 		return publicKey;
 	}
+	
+	protected Boolean checkRole(DecodedJWT jwt, String role) {
```
Comment (hankwallace) [inline src/main/java/ca/uhn/fhir/jpa/starter/OAuth2Helper.java]: Since this is returning a `Boolean`, I would expect it to be called something like `hasClientRole`. A signature of `Boolean hasClientRole(DecodedJWT jwt, String clientId, String roleName)` would be better.

**Review by hankwallace — CHANGES_REQUESTED** [review]

File: src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java [hunk 5]
```diff
@@ -29,6 +29,7 @@
 	private static final String APIKEY_HEADER = "x-api-key";
 	private static final String APIKEY = System.getenv("APIKEY");
```
Comment (hankwallace) [inline src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java:30]: Please update the `docker-compose.yml` to include the env var changes - it still has the old env var names. It would be good for it to include defaults for the `APIKEY*` and `OAUTH*` settings.

**Review by hankwallace — CHANGES_REQUESTED** [review]

File: docker-compose.yml [hunk 6]
```diff
@@ -5,19 +5,25 @@ services:
...
-      HAPI_FHIR_VERSION: 'R4'
-      HAPI_SERVER_ADDRESS: 'http://localhost:8080/fhir/'
-      HAPI_DATASOURCE_DRIVER: 'com.mysql.jdbc.Driver'
-      HAPI_DATASOURCE_URL: 'jdbc:mysql://localhost:3306/hapi'
-      HAPI_DATASOURCE_USERNAME: 'admin'
-      HAPI_DATASOURCE_PASSWORD: 'admin'
-      HAPI_HIBERNATE_DIALECT: 'org.hibernate.dialect.MySQL5InnoDBDialect'
-      API_KEY: '<apiKey>'
-      VALIDATE_API_KEY: 'false'
+      APIKEY: '<apiKey>'
```
Comment (hankwallace) [inline docker-compose.yml]: Please see if things will work if `APIKEY*` settings are not here. The `APIKEY_ENABLED` defaults to `false`, right?

File: docker-compose.yml [hunk 7]
```diff
@@ -5,19 +5,25 @@ services:
...
+      APIKEY_ENABLED: 'false'
+      datasource.driver: 'com.mysql.jdbc.Driver'
+      datasource.password: 'admin'
+      datasource.url: 'jdbc:mysql://localhost:3306/hapi'
+      datasource.username: 'admin'
+      enable_web: 'true'
+      fhir_version: 'R4'
+      hapi.fhir.server_address: 'http://localhost:8080/fhir/'
+      hibernate.dialect: 'org.hibernate.dialect.MySQL5InnoDBDialect'
+      OAUTH_CLIENT_ID: '<oauth_clientId>'
```
Comment (hankwallace) [inline docker-compose.yml]: Same for `OAUTH*`, right? Does the `OAUTH_ENABLED` default to `false`?

**Review by hankwallace — CHANGES_REQUESTED** [review]
Please remove the settings that will work if the defaults are used.

File: docker-compose.yml [hunk 8]
```diff
@@ -5,19 +5,18 @@ services:
     build: .
     container_name: hapi-fhir-jpaserver-start
     restart: on-failure
-    ports:
```
Comment (hankwallace) [inline docker-compose.yml:8]: Why did you remove the `ports` section that was in the upstream repo?

File: docker-compose.yml [hunk 9]
```diff
@@ -5,19 +5,18 @@ services:
...
-      - '8080:8080'
     environment:
-      JAVA_OPTS: '-Dhapi.properties=/usr/local/tomcat/webapps/ROOT/WEB-INF/classes/hapi-r4.properties -Denable-web=false -Durl_pattern=/fhir/*'
```
Comment (hankwallace) [inline docker-compose.yml:11]: I asked for you to remove anything that was set to the default values, but we should leave in the things that are in the upstream repo to limit our changes and make it easier to apply their updates.

File: src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java [hunk 10]
```diff
@@ -91,7 +93,10 @@
 			publicKey = StringUtils.isEmpty(publicKey) ? oAuth2Helper.getJwtPublicKey(kid, OAUTH_URL) : publicKey;
 			JWTVerifier verifier = oAuth2Helper.getJWTVerifier(jwt, publicKey);
 			jwt = verifier.verify(token);
-			return allowAll();
+			
+			if (oAuth2Helper.hasClientRole(jwt, OAUTH_USER_ROLE, OAUTH_CLIENT_ID)) {
```
Comment (hankwallace) [inline src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java]: Please switch the order of these arguments. I would expect `hasClientRole` would have the client argument before the role argument.

**Review by hankwallace — CHANGES_REQUESTED** [review]
Looks better now. A couple more suggestions.

File: docker-compose.yml (hunk 9 above)
Comment (shubhamparikh927) [reply docker-compose.yml:11]: In upstream repo also JAVA_OPTS are removed.

Comment (hankwallace) [reply docker-compose.yml:11]: That's fine. What about the lines after this (i.e. the `HAPI_FHIR*` vars)?

**Review by hankwallace — APPROVED** [review]
